### PR TITLE
precompile the app tree (to remove stuff like decorators)

### PR DESCRIPTION
**Pull request by kellyselden**
needs a test

### CODE REVIEW

File: lib/broccoli/ember-app.js [hunk 1]
```diff
@@ -760,6 +762,36 @@ class EmberApp {
     return tree;
   }
 
+  _precompileAppJsTree(tree) {
```
Comment (twokul) [inline lib/broccoli/ember-app.js]: conceptually, we should move _all_ transpilation from `ember-app` and defer that to the packager. can you move this functionality to the packager?

Comment (kellyselden) [reply lib/broccoli/ember-app.js]: We need to provide the package hook ES current, therefore this needs to be before the packager.

File: lib/broccoli/ember-app.js [hunk 2]
```diff
@@ -974,7 +1006,7 @@ class EmberApp {
    * @method getAppJavascript
    * @return {BroccoliTree}
   */
-  getAppJavascript() {
+  getAppJavascript(isPackageHookSupplied) {
```
Comment (kellyselden) [inline lib/broccoli/ember-app.js]: @twokul I added the optimization you asked for.

File: lib/broccoli/ember-app.js [hunk 3]
```diff
@@ -985,11 +1017,17 @@ class EmberApp {
...
     });
 
-    return new Funnel(mergedApp, {
+    let appTree = new Funnel(mergedApp, {
       srcDir: '/',
       destDir: this.name,
       annotation: 'ProcessedAppTree',
     });
+
+    if (experiments.DELAYED_TRANSPILATION && isPackageHookSupplied) {
```
Comment (twokul) [inline lib/broccoli/ember-app.js]: shouldn't we check if packager experiment is enabled or not rather than checking for `package` property?

**Pull request merged**